pkg/providers/cloudstack/cloudstack.go

```diff
@@ -232,25 +232,16 @@ type ProviderKubectlClient interface {
 }
 
 // NewProvider initializes the CloudStack provider object.
-func NewProvider(datacenterConfig *v1alpha1.CloudStackDatacenterConfig, machineConfigs map[string]*v1alpha1.CloudStackMachineConfig, clusterConfig *v1alpha1.Cluster, providerKubectlClient ProviderKubectlClient, validator ProviderValidator, writer filewriter.FileWriter, now types.NowFunc, log logr.Logger) *cloudstackProvider { //nolint:revive
-	controlPlaneMachineSpec := getControlPlaneMachineSpec(clusterConfig.Spec, machineConfigs)
-	etcdMachineSpec := getEtcdMachineSpec(clusterConfig.Spec, machineConfigs)
-	workerNodeGroupMachineSpecs := make(map[string]v1alpha1.CloudStackMachineConfigSpec, len(machineConfigs))
+func NewProvider(datacenterConfig *v1alpha1.CloudStackDatacenterConfig, clusterConfig *v1alpha1.Cluster, providerKubectlClient ProviderKubectlClient, validator ProviderValidator, writer filewriter.FileWriter, now types.NowFunc, log logr.Logger) *cloudstackProvider { //nolint:revive
 	return &cloudstackProvider{
 		datacenterConfig:      datacenterConfig,
-		machineConfigs:        machineConfigs,
```

Comment:
why taking off machine configs here?

Reply:
The same thing was done for the vSphere implementation a while back, of which cloudstack was essentially similar.

It's already in the` cluster.Spec` which is already passed and used in most methods, so it isn't needed. I had run into issues where updates were here but not in the cluster.Spec. Removing the `cloudstackProvider` dependency on API machineconfig structs avoids those bugs where updates are only performed in one place. Given the cluster spec is passed to most methods as an arguments, this is easier to track as the single source of truth.

Reply:
sorry let me rephrase my question: why taking of machine config but leave datacenter? Could you check if datacenter can be removed as well? not necessarily in this pr if you think that involves too much work, but can create an issue to track

Reply:
It think we should do this in another PR, as a fast follow. It won't be too much work, but testing may take up some more time. Also, I do not want to extend the scope of this PR further.